Added Node/npm version plugin

castellanprime opened the pull request:
This is version 1. it is still missing the enhancements talked out in the last commit. #57

castellanprime committed:
Created scaffolding file

castellanprime committed:
Added node to GEOMEMTRY_PLUGINS variable

castellanprime committed:
Made the information print on the terminal

castellanprime committed:
Added a README.md in the mold of the hg plugin README.md

castellanprime committed:
Made slight modification to README.md: Added color black

castellanprime committed:
Prints the node and npm versions as given: node version (npm version)

	* Bug: Can only print character symbols not the given symbol

	* Bug: Prints on initial login.

	* Enhancement: Print only after checking for yarn/nvm/nodeenv installation

desyncr commented:
Looks great to me :+1:. Just gonna add a few comments.

desyncr commented on lib/plugin.zsh:
```diff
@@ -4,7 +4,7 @@ typeset -g GEOMETRY_PLUGIN_SEPARATOR=${GEOMETRY_PLUGIN_SEPARATOR:-" "}
 typeset -ga GEOMETRY_PROMPT_PLUGINS
 if [[ $#GEOMETRY_PROMPT_PLUGINS -eq 0 ]]; then
-  GEOMETRY_PROMPT_PLUGINS=(exec_time git hg)
+  GEOMETRY_PROMPT_PLUGINS=(exec_time git hg node)
```
Not sure about adding it as a default built-in plugin. @frmendes What do you think?

castellanprime replied:
I will remove it from the default built-in plugin.

frm replied:
I'd rather not have it as a default plugin, to keep the "minimal default" philosophy

desyncr commented on plugins/node/README.md:
````diff
@@ -0,0 +1,18 @@
+### Colors
+
+```sh
+GEOMETRY_COLOR_NODE_NPM_VERISON="black"
````
Typo: `VERISON`  -> `VERSION`.

castellanprime replied:
Okay.

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,18 @@
+# Color definitions
+GEOMETRY_COLOR_NODE_NPM_VERISON=${GEOMETRY_COLOR_NODE_NPM_VERISON:-black}
```
Typo same as above.

castellanprime replied:
Will change

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,18 @@
+GEOMETRY_SYMBOL_NODE_NPM_VERISON=${GEOMETRY_SYMBOL_NODE_NPM_VERSION:-"⬡"}
+GEOMETRY_NODE_NPM_VERISON=$(prompt_geometry_colorize $GEOMETRY_COLOR_NODE_NPM_VERISON GEOMETRY_SYMBOL_NODE_NPM_VERISON) 
+
+geometry_prompt_node_setup() 
```
Should be `geometry_prompt_node_setup() {}` in order to define the function correctly.

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,18 @@
+geometry_prompt_node_render(){
+	local node_path="$echo "$(which node)" -v"
+	local npm_path="$echo "$(which npm)" --version"
+	GEOMETRY_NODE_VERSION=`(eval ${node_path})`
```
There are a couple of issues here but my suggestion is to keep it simple:

1- Check node/npm are available
2- Display version info

For 1) zsh has a neat way to do it:

`$+commands[node]` (or for `npm`: `$+commands[npm]`).


You should check if there is no node/npm command available and return an empty string if none is available:

    if (( ! $+commands[node] )); then
        return "";
    fi


For 2) you can simply do:

    GEOMETRY_NODE_VERSION="$(node -v)"

castellanprime replied:
From your explanation, the statement "$+commands[node]" returns an alias "node". Is this right? If so, how to add npm to this array?

desyncr replied:
`$+commands[node]` will return `1` if it finds `node` in the `commands` array. Example:

With `node` command available in `$PATH`:

    > echo $+commands[node]
    1

Without `node` available:

    > echo $+commands[node]
    0

If node is found in `$+commands` we should be able to do `node -v`, otherwise I'd return immediately without displaying anything.

desyncr reviewed (changes requested):
There are a couple of issues but great work so far! :)

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,18 @@
+
+# Symbol definitions
+GEOMETRY_SYMBOL_NODE_NPM_VERISON=${GEOMETRY_SYMBOL_NODE_NPM_VERSION:-"⬡"}
+GEOMETRY_NODE_NPM_VERISON=$(prompt_geometry_colorize $GEOMETRY_COLOR_NODE_NPM_VERISON GEOMETRY_SYMBOL_NODE_NPM_VERISON) 
```
Here you are missing `$` on `GEOMETRY_SYMBOL_NODE_NPM_VERISON`. That's the reason the symbol is not displayed.

castellanprime replied:
Thanks. Will change now.

desyncr reviewed (changes requested):
One more comment :)

castellanprime commented:
I am sorry that I have not committed anything towards this in the last few days. I have been very busy. I will work on the code from tonight.

desyncr commented:
@castellanprime awesome. Don't worry. Keep up the good work!

castellanprime committed:
Fixed typo VERISON->VERSION

castellanprime committed:
Fixed typos and added changes that was requested

	* Fixed typo VERISON-> VERSION

	* Used commands array to check for the occurence of node on the system

	* Bugs: Still yet to check for yarn/nvm installation

castellanprime committed:
Removed node from the default plugin list

castellanprime commented:
I have fixed the errors in the pull request.

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,24 @@
+GEOMETRY_SYMBOL_NODE_NPM_VERSION=${GEOMETRY_SYMBOL_NODE_NPM_VERSION:-"⬡"}
+GEOMETRY_NODE_NPM_VERSION=$(prompt_geometry_colorize $GEOMETRY_COLOR_NODE_NPM_VERSION $GEOMETRY_SYMBOL_NODE_NPM_VERSION) 
+
+local node='node'
```
No need to put it into a variable.

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,24 @@
+geometry_prompt_node_setup(){} 
+
+geometry_prompt_node_render(){
+	local n=$+commands[$node]
```
Same here. No need to use a variable. Better  use:

    if (( $+commands ['node'] == 0 ))

desyncr reviewed (changes requested):
Great! Just a few changes.

Also use 4 spaces for indentation.

castellanprime committed:
Convert to tabs to spaces and remove temporary variable

castellanprime commented:
Added your changes

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,21 @@
+geometry_prompt_node_setup(){} 
+
+geometry_prompt_node_render(){
+    if (( $+commands ['node'] == 0 )); then
```
I think I gave you a bad tip. This line should read as follow:

    if (( $+commands[node] == 0 )); then

So, instead of `'node'` (quoted), it should be `node`.

castellanprime replied:
Fixed that. Just pushed. Thanks for your help so far.

castellanprime replied:
I am working on the detecting the yarn.lock file and .nvmrc file but I am running into problems. This is the line I want to add:
`if [[-e yarn.lock]] || [[ -f .nmvrc && -r .nvrmc && -n "$(head -1 .nvmrc)]]; then"`.
I want add this above the 
`if (( $+commands[node] == 0)); then`

I tested it using a simple .nvmrc file in a directory jsproj using this steps 
![screenshot from 2017-01-23 22 58 29](https://cloud.githubusercontent.com/assets/15849318/22234701/b82296dc-e1bf-11e6-8ff8-276956856d01.png)

desyncr reviewed (changes requested):
I gave you a wrong example. My bad.

This small fix up and we'll be done and ready! Cheers!

castellanprime committed:
Fixed the typo in 0 -> 1

castellanprime committed:
Modified README.md

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,21 @@
+geometry_prompt_node_setup(){} 
+
+geometry_prompt_node_render(){	
+    if (( $+commands[node] == 0 )); then
```
Remember that `$+commands[node]` will return `0` if the command does not exists, `1` otherwise. So, this conditional is backwards.

desyncr commented on plugins/node/plugin.zsh:
```diff
@@ -0,0 +1,21 @@
+        GEOMETRY_NPM_VERSION="$(npm --version)" 
+        echo "$GEOMETRY_NODE_NPM_VERSION $GEOMETRY_NODE_VERSION ($GEOMETRY_NPM_VERSION)"
+    else
+        echo " "
```
In case there is no node available do not generate an empty space, ie: `echo ""`.

desyncr reviewed (changes requested):
Getting closer. Review the conditional and an empty space and we'll be done!

castellanprime committed:
Removed unneccessary conditional and modified main conditional

castellanprime commented:
Added your changes. I wonder if you looked at my earlier suggestion about the .nmvrc/yarn.lock files?  Is that what you wanted? Thanks.

desyncr reviewed (approved):
Very happy with your work! Gonna merge this now!

desyncr commented:
@castellanprime Regarding `yarn.lock` I think it's a great addition to put us ahead of other prompts. If you are up for it we can work on this on another PR. :)
